Exit tests crash main process on Windows (testing fails)

### Description

When enabling exit tests on Windows, the tests fail indicating that `swift.exe` unexpectedly exited:

```
Error: Process completed with exit code 1.
```

### Reproduction

We're seeing this with any exit test we enable on Windows, to reproduce simply create a test that uses `#expect(processExitsWith: .failure)` and testing will fail with this message

### Expected behavior

I'd expect tests to succeed and not crash (except for the expected crashing code) on Windows like the other platforms

### Environment

https://github.com/swiftlang/swift-foundation/actions/runs/16126481664/job/45504492458?pr=1406

```
Swift version 6.2-dev (LLVM 3c4a54b02dd62b7, Swift a40a7be6945a895)
Target: x86_64-unknown-windows-msvc
Build config: +assertions
```

```
→ Testing Library Version: 6.3-dev (25b61ef4667a6d1)
```

### Additional information

Note that the same exit tests in https://github.com/swiftlang/swift-foundation/actions/runs/16126481664?pr=1406 pass on Linux and macOS but hit this issue on Windows

--- a/Sources/Testing/ExitTests/SpawnProcess.swift
+++ b/Sources/Testing/ExitTests/SpawnProcess.swift
@@ -297,9 +297,9 @@ func spawnExecutable(
       //
       // SEE: https://devblogs.microsoft.com/oldnewthing/20101109-00/?p=12323
       let workingDirectoryPath: UnsafeMutablePointer<wchar_t>? = {
-        let systemDrive = Environment.variable(named: "SYSTEMDRIVE") ?? "C:"
+        var systemDrive = Environment.variable(named: "SYSTEMDRIVE") ?? "C:"
         if systemDrive.last == ":" {
-          return #"\#(systemDrive)\"#
+          systemDrive = #"\#(systemDrive)\"#
         }
         return systemDrive.withCString(encodedAs: UTF16.self) { _wcsdup($0) }
       }()

--- a/Sources/Testing/ExitTests/SpawnProcess.swift
+++ b/Sources/Testing/ExitTests/SpawnProcess.swift
@@ -284,6 +284,29 @@ func spawnExecutable(
       let commandLine = _escapeCommandLine(CollectionOfOne(executablePath) + arguments)
       let environ = environment.map { "\($0.key)=\($0.value)" }.joined(separator: "\0") + "\0\0"
 
+      // On Windows, a process holds a reference to its current working
+      // directory, which prevents other processes from deleting it. This causes
+      // code to fail if it tries to set the working directory to a temporary
+      // path. SEE: https://github.com/swiftlang/swift-testing/issues/1209
+      //
+      // This problem manifests for us when we spawn a child process without
+      // setting its working directory, which causes it to default to that of
+      // the parent process. To avoid this problem, we set the working directory
+      // of the new process to the root directory of the boot volume (which is
+      // unlikely to be deleted, one hopes).
+      //
+      // SEE: https://devblogs.microsoft.com/oldnewthing/20101109-00/?p=12323
+      let workingDirectoryPath: UnsafeMutablePointer<wchar_t>? = {
+        let systemDrive = Environment.variable(named: "SYSTEMDRIVE") ?? "C:"
+        if systemDrive.last == ":" {
+          return #"\#(systemDrive)\"#
+        }
+        return systemDrive.withCString(encodedAs: UTF16.self) { _wcsdup($0) }
+      }()
+      defer {
+        free(workingDirectoryPath)
+      }
+
       var flags = DWORD(CREATE_NO_WINDOW | CREATE_UNICODE_ENVIRONMENT | EXTENDED_STARTUPINFO_PRESENT)
 #if DEBUG
       // Start the process suspended so we can attach a debugger if needed.
@@ -302,7 +325,7 @@ func spawnExecutable(
             true, // bInheritHandles
             flags,
             .init(mutating: environ),
-            nil,
+            workingDirectoryPath,
             startupInfo.pointer(to: \.StartupInfo)!,
             &processInfo
           ) else {

--- a/Sources/Testing/ExitTests/SpawnProcess.swift
+++ b/Sources/Testing/ExitTests/SpawnProcess.swift
@@ -229,7 +229,6 @@ func spawnExecutable(
       return pid
     }
   }
-
 #elseif os(Windows)
   return try _withStartupInfoEx(attributeCount: 1) { startupInfo in
     func inherit(_ fileHandle: borrowing FileHandle) throws -> HANDLE? {
@@ -285,6 +284,14 @@ func spawnExecutable(
       let commandLine = _escapeCommandLine(CollectionOfOne(executablePath) + arguments)
       let environ = environment.map { "\($0.key)=\($0.value)" }.joined(separator: "\0") + "\0\0"
 
+      // CreateProcessW() may modify the command line argument, so we must make
+      // a mutable copy of it. (environ is also passed as a mutable raw pointer,
+      // but it is not documented as actually being mutated.)
+      let commandLineCopy = commandLine.withCString(encodedAs: UTF16.self) { _wcsdup($0) }
+      defer {
+        free(commandLineCopy)
+      }
+
       // On Windows, a process holds a reference to its current working
       // directory, which prevents other processes from deleting it. This causes
       // code to fail if it tries to set the working directory to a temporary
@@ -305,34 +312,32 @@ func spawnExecutable(
       flags |= DWORD(CREATE_SUSPENDED)
 #endif
 
-      return try commandLine.withCString(encodedAs: UTF16.self) { commandLine in
-        try environ.withCString(encodedAs: UTF16.self) { environ in
-          try workingDirectoryPath.withCString(encodedAs: UTF16.self) { workingDirectoryPath in
-            var processInfo = PROCESS_INFORMATION()
+      return try environ.withCString(encodedAs: UTF16.self) { environ in
+        try workingDirectoryPath.withCString(encodedAs: UTF16.self) { workingDirectoryPath in
+          var processInfo = PROCESS_INFORMATION()
 
-            guard CreateProcessW(
-              nil,
-              .init(mutating: commandLine),
-              nil,
-              nil,
-              true, // bInheritHandles
-              flags,
-              .init(mutating: environ),
-              workingDirectoryPath,
-              startupInfo.pointer(to: \.StartupInfo)!,
-              &processInfo
-            ) else {
-              throw Win32Error(rawValue: GetLastError())
-            }
+          guard CreateProcessW(
+            nil,
+            commandLineCopy,
+            nil,
+            nil,
+            true, // bInheritHandles
+            flags,
+            .init(mutating: environ),
+            workingDirectoryPath,
+            startupInfo.pointer(to: \.StartupInfo)!,
+            &processInfo
+          ) else {
+            throw Win32Error(rawValue: GetLastError())
+          }
 
 #if DEBUG
-            // Resume the process.
-            _ = ResumeThread(processInfo.hThread!)
+          // Resume the process.
+          _ = ResumeThread(processInfo.hThread!)
 #endif
 
-            _ = CloseHandle(processInfo.hThread)
-            return processInfo.hProcess!
-          }
+          _ = CloseHandle(processInfo.hThread)
+          return processInfo.hProcess!
         }
       }
     }

--- a/Sources/Testing/ExitTests/SpawnProcess.swift
+++ b/Sources/Testing/ExitTests/SpawnProcess.swift
@@ -297,18 +297,7 @@ func spawnExecutable(
       // unlikely to be deleted, one hopes).
       //
       // SEE: https://devblogs.microsoft.com/oldnewthing/20101109-00/?p=12323
-      var workingDirectoryPath = UnsafeMutableBufferPointer<wchar_t>.allocate(capacity: Int(MAX_PATH))
-      defer {
-        workingDirectoryPath.deallocate()
-      }
-      let systemDrive = Environment.variable(named: "SYSTEMDRIVE") ?? "C:"
-      systemDrive.withCString(encodedAs: UTF16.self) { systemDrive in
-        wcscpy_s(workingDirectoryPath.baseAddress!, workingDirectoryPath.count, systemDrive)
-        let rAddBackslash = PathCchAddBackslashEx(workingDirectoryPath.baseAddress!, workingDirectoryPath.count, nil, nil)
-        guard rAddBackslash == S_OK || rAddBackslash == S_FALSE else {
-          fatalError("Unexpected error when normalizing system drive path '\(systemDrive)': HRESULT(\(rAddBackslash))")
-        }
-      }
+      let workingDirectoryPath = rootDirectoryPath
 
       var flags = DWORD(CREATE_NO_WINDOW | CREATE_UNICODE_ENVIRONMENT | EXTENDED_STARTUPINFO_PRESENT)
 #if DEBUG
@@ -318,30 +307,32 @@ func spawnExecutable(
 
       return try commandLine.withCString(encodedAs: UTF16.self) { commandLine in
         try environ.withCString(encodedAs: UTF16.self) { environ in
-          var processInfo = PROCESS_INFORMATION()
+          try workingDirectoryPath.withCString(encodedAs: UTF16.self) { workingDirectoryPath in
+            var processInfo = PROCESS_INFORMATION()
 
-          guard CreateProcessW(
-            nil,
-            .init(mutating: commandLine),
-            nil,
-            nil,
-            true, // bInheritHandles
-            flags,
-            .init(mutating: environ),
-            workingDirectoryPath,
-            startupInfo.pointer(to: \.StartupInfo)!,
-            &processInfo
-          ) else {
-            throw Win32Error(rawValue: GetLastError())
-          }
+            guard CreateProcessW(
+              nil,
+              .init(mutating: commandLine),
+              nil,
+              nil,
+              true, // bInheritHandles
+              flags,
+              .init(mutating: environ),
+              workingDirectoryPath,
+              startupInfo.pointer(to: \.StartupInfo)!,
+              &processInfo
+            ) else {
+              throw Win32Error(rawValue: GetLastError())
+            }
 
 #if DEBUG
-          // Resume the process.
-          _ = ResumeThread(processInfo.hThread!)
+            // Resume the process.
+            _ = ResumeThread(processInfo.hThread!)
 #endif
 
-          _ = CloseHandle(processInfo.hThread)
-          return processInfo.hProcess!
+            _ = CloseHandle(processInfo.hThread)
+            return processInfo.hProcess!
+          }
         }
       }
     }

--- a/Sources/Testing/ExitTests/SpawnProcess.swift
+++ b/Sources/Testing/ExitTests/SpawnProcess.swift
@@ -229,6 +229,7 @@ func spawnExecutable(
       return pid
     }
   }
+
 #elseif os(Windows)
   return try _withStartupInfoEx(attributeCount: 1) { startupInfo in
     func inherit(_ fileHandle: borrowing FileHandle) throws -> HANDLE? {
@@ -296,15 +297,17 @@ func spawnExecutable(
       // unlikely to be deleted, one hopes).
       //
       // SEE: https://devblogs.microsoft.com/oldnewthing/20101109-00/?p=12323
-      let workingDirectoryPath: UnsafeMutablePointer<wchar_t>? = {
-        var systemDrive = Environment.variable(named: "SYSTEMDRIVE") ?? "C:"
-        if systemDrive.last == ":" {
-          systemDrive = #"\#(systemDrive)\"#
-        }
-        return systemDrive.withCString(encodedAs: UTF16.self) { _wcsdup($0) }
-      }()
+      var workingDirectoryPath = UnsafeMutableBufferPointer<wchar_t>.allocate(capacity: Int(MAX_PATH))
       defer {
-        free(workingDirectoryPath)
+        workingDirectoryPath.deallocate()
+      }
+      let systemDrive = Environment.variable(named: "SYSTEMDRIVE") ?? "C:"
+      systemDrive.withCString(encodedAs: UTF16.self) { systemDrive in
+        wcscpy_s(workingDirectoryPath.baseAddress!, workingDirectoryPath.count, systemDrive)
+        let rAddBackslash = PathCchAddBackslashEx(workingDirectoryPath.baseAddress!, workingDirectoryPath.count, nil, nil)
+        guard rAddBackslash == S_OK || rAddBackslash == S_FALSE else {
+          fatalError("Unexpected error when normalizing system drive path '\(systemDrive)': HRESULT(\(rAddBackslash))")
+        }
       }
 
       var flags = DWORD(CREATE_NO_WINDOW | CREATE_UNICODE_ENVIRONMENT | EXTENDED_STARTUPINFO_PRESENT)

--- a/Sources/Testing/Support/FileHandle.swift
+++ b/Sources/Testing/Support/FileHandle.swift
@@ -741,11 +741,11 @@ let rootDirectoryPath: String = {
         result = String.decodeCString(buffer.baseAddress!, as: UTF16.self)?.result
       }
     }
-
-    // If we weren't able to get a path, fall back to "C:\" on the assumption
-    // that it's the common case and most likely correct.
-    return result ?? #"C:\"#
   }
+
+  // If we weren't able to get a path, fall back to "C:\" on the assumption that
+  // it's the common case and most likely correct.
+  return result ?? #"C:\"#
 #else
   return "/"
 #endif

--- a/Sources/Testing/Support/FileHandle.swift
+++ b/Sources/Testing/Support/FileHandle.swift
@@ -719,4 +719,37 @@ func setFD_CLOEXEC(_ flag: Bool, onFileDescriptor fd: CInt) throws {
   }
 }
 #endif
+
+/// The path to the root directory of the boot volume.
+///
+/// On Windows, this string is usually of the form `"C:\"`. On UNIX-like
+/// platforms, it is always equal to `"/"`.
+let rootDirectoryPath: String = {
+#if os(Windows)
+  if let systemDrive = Environment.variable(named: "SYSTEMDRIVE")?.utf16 {
+    // Copy the system drive string with room for up to "C:\" and a NUL.
+    var buffer = UnsafeMutableBufferPointer<wchar_t>.allocate(capacity: systemDrive.count + 4)
+    defer {
+      buffer.deallocate()
+    }
+    buffer.initialize(fromContentsOf: systemDrive)
+    buffer[systemDrive.count] = 0
+
+    // On the assumption that the value of %SYSTEMDRIVE% is "C:" or similar,
+    // ensure a trailing slash is added to refer to the root directory. If
+    // somebody decides to set this environment variable to something
+    // nonsensical or to a deeper path, we should accept it silently.
+    let rAddBackslash = PathCchAddBackslashEx(buffer.baseAddress!, buffer.count, nil, nil)
+    if rAddBackslash == S_OK || rAddBackslash == S_FALSE,
+       let result = String.decodeCString(buffer.baseAddress!, as: UTF16.self)?.result {
+      return result
+    }
+  }
+  // We weren't able to get a path, so fall back to "C:\" on the assumption that
+  // it's the common case and most likely correct.
+  return #"C:\"#
+#else
+  return "/"
+#endif
+}()
 #endif

--- a/Sources/Testing/Support/FileHandle.swift
+++ b/Sources/Testing/Support/FileHandle.swift
@@ -726,28 +726,26 @@ func setFD_CLOEXEC(_ flag: Bool, onFileDescriptor fd: CInt) throws {
 /// platforms, it is always equal to `"/"`.
 let rootDirectoryPath: String = {
 #if os(Windows)
-  if let systemDrive = Environment.variable(named: "SYSTEMDRIVE")?.utf16 {
-    // Copy the system drive string with room for up to "C:\" and a NUL.
-    var buffer = UnsafeMutableBufferPointer<wchar_t>.allocate(capacity: systemDrive.count + 4)
-    defer {
-      buffer.deallocate()
-    }
-    buffer.initialize(fromContentsOf: systemDrive)
-    buffer[systemDrive.count] = 0
+  var result: String?
 
-    // On the assumption that the value of %SYSTEMDRIVE% is "C:" or similar,
-    // ensure a trailing slash is added to refer to the root directory. If
-    // somebody decides to set this environment variable to something
-    // nonsensical or to a deeper path, we should accept it silently.
-    let rAddBackslash = PathCchAddBackslashEx(buffer.baseAddress!, buffer.count, nil, nil)
-    if rAddBackslash == S_OK || rAddBackslash == S_FALSE,
-       let result = String.decodeCString(buffer.baseAddress!, as: UTF16.self)?.result {
-      return result
+  // The boot volume is, except in some legacy scenarios, the volume that
+  // contains the system Windows directory. For an explanation of the difference
+  // between the Windows directory and the _system_ Windows directory, see
+  // https://devblogs.microsoft.com/oldnewthing/20140723-00/?p=423 .
+  let count = GetSystemWindowsDirectoryW(nil, 0)
+  if count > 0 {
+    withUnsafeTemporaryAllocation(of: wchar_t.self, capacity: Int(count) + 1) { buffer in
+      _ = GetSystemWindowsDirectoryW(buffer.baseAddress!, UINT(buffer.count))
+      let rStrip = PathCchStripToRoot(buffer.baseAddress!, buffer.count)
+      if rStrip == S_OK || rStrip == S_FALSE {
+        result = String.decodeCString(buffer.baseAddress!, as: UTF16.self)?.result
+      }
     }
+
+    // If we weren't able to get a path, fall back to "C:\" on the assumption
+    // that it's the common case and most likely correct.
+    return result ?? #"C:\"#
   }
-  // We weren't able to get a path, so fall back to "C:\" on the assumption that
-  // it's the common case and most likely correct.
-  return #"C:\"#
 #else
   return "/"
 #endif

--- a/Tests/TestingTests/Support/FileHandleTests.swift
+++ b/Tests/TestingTests/Support/FileHandleTests.swift
@@ -213,7 +213,7 @@ struct FileHandleTests {
       #expect(Environment.setVariable("Q:", named: "SYSTEMDRIVE"))
       #expect(rootDirectoryPath == #"Q:\"#)
 
-      #expect(Environment.setVariable("Q:\abc123", named: "SYSTEMDRIVE"))
+      #expect(Environment.setVariable(#"Q:\abc123"#, named: "SYSTEMDRIVE"))
       #expect(rootDirectoryPath == #"Q:\abc123\"#)
     }
 #endif

--- a/Tests/TestingTests/Support/FileHandleTests.swift
+++ b/Tests/TestingTests/Support/FileHandleTests.swift
@@ -201,6 +201,26 @@ struct FileHandleTests {
 #endif
   }
 #endif
+
+  @Test("Root directory path is correct")
+  func rootDirectoryPathIsCorrect() async {
+#if os(Windows)
+#if !SWT_NO_EXIT_TESTS
+    await #expect(processExitsWith: .success) {
+      #expect(Environment.setVariable(nil, named: "SYSTEMDRIVE"))
+      #expect(rootDirectoryPath == #"C:\"#)
+
+      #expect(Environment.setVariable("Q:", named: "SYSTEMDRIVE"))
+      #expect(rootDirectoryPath == #"Q:\"#)
+
+      #expect(Environment.setVariable("Q:\abc123", named: "SYSTEMDRIVE"))
+      #expect(rootDirectoryPath == #"Q:\abc123\"#)
+    }
+#endif
+#else
+    #expect(rootDirectoryPath == "/")
+#endif
+  }
 }
 
 // MARK: - Fixtures

--- a/Tests/TestingTests/Support/FileHandleTests.swift
+++ b/Tests/TestingTests/Support/FileHandleTests.swift
@@ -203,24 +203,11 @@ struct FileHandleTests {
 #endif
 
   @Test("Root directory path is correct")
-  func rootDirectoryPathIsCorrect() async {
+  func rootDirectoryPathIsCorrect() throws {
 #if os(Windows)
-#if !SWT_NO_EXIT_TESTS
-    await #expect(processExitsWith: .success) {
-      #expect(Environment.setVariable(nil, named: "SYSTEMDRIVE"))
-      #expect(rootDirectoryPath == #"C:\"#)
+    if let systemDrive = Environment.variable(named: "SYSTEMDRIVE") {
+      #expect(rootDirectoryPath.starts(with: systemDrive))
     }
-
-    await #expect(processExitsWith: .success) {
-      #expect(Environment.setVariable("Q:", named: "SYSTEMDRIVE"))
-      #expect(rootDirectoryPath == #"Q:\"#)
-    }
-
-    await #expect(processExitsWith: .success) {
-      #expect(Environment.setVariable(#"Q:\abc123"#, named: "SYSTEMDRIVE"))
-      #expect(rootDirectoryPath == #"Q:\abc123\"#)
-    }
-#endif
 #else
     #expect(rootDirectoryPath == "/")
 #endif

--- a/Tests/TestingTests/Support/FileHandleTests.swift
+++ b/Tests/TestingTests/Support/FileHandleTests.swift
@@ -209,10 +209,14 @@ struct FileHandleTests {
     await #expect(processExitsWith: .success) {
       #expect(Environment.setVariable(nil, named: "SYSTEMDRIVE"))
       #expect(rootDirectoryPath == #"C:\"#)
+    }
 
+    await #expect(processExitsWith: .success) {
       #expect(Environment.setVariable("Q:", named: "SYSTEMDRIVE"))
       #expect(rootDirectoryPath == #"Q:\"#)
+    }
 
+    await #expect(processExitsWith: .success) {
       #expect(Environment.setVariable(#"Q:\abc123"#, named: "SYSTEMDRIVE"))
       #expect(rootDirectoryPath == #"Q:\abc123\"#)
     }

